src/coreclr/src/gc/gc.cpp

```diff
@@ -26105,65 +28189,75 @@ void gc_heap::update_oldest_pinned_plug()
 void gc_heap::relocate_survivors (int condemned_gen_number,
                                   uint8_t* first_condemned_address)
 {
-    generation* condemned_gen = generation_of (condemned_gen_number);
-    uint8_t*  start_address = first_condemned_address;
-    size_t  current_brick = brick_of (start_address);
-    heap_segment*  current_heap_segment = heap_segment_rw (generation_start_segment (condemned_gen));
-
-    PREFIX_ASSUME(current_heap_segment != NULL);
-
-    uint8_t*  end_address = 0;
-
     reset_pinned_queue_bos();
     update_oldest_pinned_plug();
 
-    end_address = heap_segment_allocated (current_heap_segment);
+    int stop_gen_idx = get_stop_generation_index (condemned_gen_number);
 
-    size_t  end_brick = brick_of (end_address - 1);
-    relocate_args args;
-    args.is_shortened = FALSE;
-    args.pinned_plug_entry = 0;
-    args.last_plug = 0;
-    while (1)
+#ifndef USE_REGIONS
+    assert (first_condemned_address == generation_allocation_start (generation_of (condemned_gen_number)));
+#endif //!USE_REGIONS
+
+    for (int i = condemned_gen_number; i >= stop_gen_idx; i--)
     {
-        if (current_brick > end_brick)
+        generation* condemned_gen = generation_of (i);
+        heap_segment*  current_heap_segment = heap_segment_rw (generation_start_segment (condemned_gen));
+        uint8_t*  start_address = get_soh_start_object (current_heap_segment, condemned_gen);
+        size_t  current_brick = brick_of (start_address);
+
+        PREFIX_ASSUME(current_heap_segment != NULL);
+
+        uint8_t*  end_address = 0;
+
+        end_address = heap_segment_allocated (current_heap_segment);
```

Comment:
Nit: you could combine this with the previous line containing the declaration - easier to read.

Reply:
will do.